### PR TITLE
HLRC support for string_stats

### PR DESCRIPTION
**Pull request by nik9000**
This adds a builder and parsed results for the `string_stats`
aggregation directly to the high level rest client. Without this the
HLRC can't access the `string_stats` API without the elastic licensed
`analytics` module.

While I'm in there this adds a few of our usual unit tests and
modernizes the parsing.

### CODE REVIEW

**Comment by nik9000** [comment]
This is a work in progress because I don't have the horsepower to run all the tests at the moment. So I'm making Jenkins do it. I *think* this is how I want to expose analytics aggs to the HLRC, though the hack with the builder is unfortunate.

**Comment by nik9000** [comment]
I've removed `WIP` because it seems to mostly work. I'm going to figure out what the test failure is this morning.

**Comment by nik9000** [comment]
@imotov you mentioned a while back that you were interested in talking about HLRC support for the analytics aggs. This is my take on explicit support for the `string_stats` agg. What do you think?

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    protected InternalStringStats mutateInstance(InternalStringStats instance) throws IOException {
+         String name = instance.getName();
+         long count = instance.getCount();
+         long totalLength = instance.getTotalLength();
+         int minLength = instance.getMinLength();
+         int maxLength = instance.getMaxLength();
+         Map<String, Long> charOccurrences = instance.getCharOccurrences();
+         boolean showDistribution = instance.getShowDistribution();
+         switch (between(0, 6)) {
+         case 0:
```
Comment (imotov) [inline x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java]: That indentation looks a bit strange to me.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/analytics/StringStatsIT.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.aMapWithSize;
+import static org.hamcrest.Matchers.closeTo;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+
+public class StringStatsIT extends ESRestHighLevelClientTestCase {
```
Comment (imotov) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/analytics/StringStatsIT.java]: Just wonder if we should rename this to `AnalyticsAggsIT` or something like this, so we can add other agg tests from analytics plugin here as we add convert them to work with the rest client.

**Review by imotov — APPROVED** [review]
Looks nice! Even though that doesn't solve an issue with `AggregationBuilders.stringStats()` method, but I think it is a good step forward.

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java]: It is the true indentation for case statements!

File: client/rest-high-level/src/test/java/org/elasticsearch/client/analytics/StringStatsIT.java (hunk 2 above)
Comment (nik9000) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/analytics/StringStatsIT.java]: Sure!

**Comment by nik9000** [comment]
> Looks nice! Even though that doesn't solve an issue with `AggregationBuilders.stringStats()` method, but I think it is a good step forward.

Do you think we should move `AggregationBuilders` into the HLRC in 8.0? And just make the aggs by hand outside of HLRC? Personally I'm not a super huge fan of the builder methods anyway, but I want to do right by them.

**Comment by nik9000** [comment]
> Do you think we should move `AggregationBuilders` into the HLRC in 8.0? And just make the aggs by hand outside of HLRC? Personally I'm not a super huge fan of the builder methods anyway, but I want to do right by them.

Like, as part of a separate change. Not this one.

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java (hunk 1 above)
Comment (imotov) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java]: It is [indeed](https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html), we just don't use it anywhere else :) that's why it looks strange.

Comment (nik9000) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStatsTests.java]: It is the Eclipse standard but not the intellij standard. I prefer this was but I'm not picky.

**Comment by imotov** [comment]
That's the challenge `*AggregationBuilder`s are playing dual role at the moment. On one side they are the client facing aggregation builders, on another side they are transport classes that know how to serialize/deserialize aggs over transport interface as well as parse them. There  are ways to separate these functions, but they are not trivial, unfortunately.

**Comment by nik9000** [comment]
> There are ways to separate these functions, but they are not trivial, unfortunately.

+2

**Comment by nik9000** [comment]
Thanks @imotov !

**Pull request merged**